### PR TITLE
Removal of module variables from ucx_mod

### PR DESCRIPTION
**Pull request by jimmielin**
This patch moves in-module variables in UCX_MOD to State_Chm, to facilitate integration of GEOS-Chem into the WRF-GC and CESM models.

This patch might be considered a draft. There are some shortcomings with this patch, namely because the UCX variables are not regularly-sized (not I, J, L), they cannot be allocated using the current implementation of Init_and_Register, and instead are allocated by ucx_mod itself. This may need an alternative implementation of State_Chm_Mod's Init_and_Register to fix.

Hopefully, this patch is mostly self-contained because most of the changes are in `ucx_mod.F90` (not very frequently updated) and only additions to `state_chm_mod.F90` (easy to merge), and little changes everywhere else. So hopefully the PR can be brought forward to future versions easily. I'm opening this up with a preliminary implementation to put this on the radar. After this, I think the other major module that still hasn't migrated to State variables is aerosol_mod.

Thanks,
Haipeng

### CODE REVIEW

**Review by msulprizio — CHANGES_REQUESTED** [review]
These changes looks good. I have just one comment below @jimmielin.

File: GeosCore/ucx_mod.F90 [hunk 1]
```diff
@@ -1537,11 +1505,16 @@ END SUBROUTINE CALC_H2SO4_GAS
...
+  REAL(fp) FUNCTION SO4_PHOTFRAC(I,J,L,State_Chm)
+!
+! !USES:
+!
+    USE State_Chm_Mod,      ONLY : ChmState
 !
 ! !INPUT PARAMETERS:
 !
     INTEGER,  INTENT(IN)  :: I,J,L      ! Location indices
+    TYPE(ChmState), INTENT(IN) :: State_Chm   ! Chemistry State object
```
[the remark this answers is not in the document]
Comment (jimmielin) [reply GeosCore/ucx_mod.F90]: Fixed! Also noticed another part where it should've been `INOUT`.

[the remark this answers is not in the document]
Comment (msulprizio) [reply GeosCore/ucx_mod.F90]: Excellent. Thanks @jimmieli! I will now approve this PR.

**Review by msulprizio — APPROVED** [review]

**Comment by yantosca** [comment]
GEOS-Chem Classic integration tests all passed:
```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

GCClassic #1915750 HEMCO submod update: Merge PR #191 into HEMCO 3.6.1
GEOS-Chem #54b11c61d Stage code for PR #1583 for merging into GEOS-Chem 14.1.1
HEMCO     #cf7d4ed GEOS-only updates by Christoph Keller

Using 24 OpenMP threads
Number of execution tests: 26

Submitted as SLURM job: 43292290
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gc_05x0625_NA_47L_merra2_CH4........................Execute Simulation....PASS
gc_05x0625_NA_47L_merra2_fullchem...................Execute Simulation....PASS
gc_4x5_47L_merra2_fullchem..........................Execute Simulation....PASS
gc_4x5_47L_merra2_fullchem_TOMAS15..................Execute Simulation....PASS
gc_4x5_47L_merra2_fullchem_TOMAS40..................Execute Simulation....PASS
gc_4x5_merra2_aerosol...............................Execute Simulation....PASS
gc_4x5_merra2_carbon................................Execute Simulation....PASS
gc_4x5_merra2_CH4...................................Execute Simulation....PASS
gc_4x5_merra2_CO2...................................Execute Simulation....PASS
gc_4x5_merra2_fullchem..............................Execute Simulation....PASS
gc_4x5_merra2_fullchem_aciduptake...................Execute Simulation....PASS
gc_4x5_merra2_fullchem_APM..........................Execute Simulation....PASS
gc_4x5_merra2_fullchem_benchmark....................Execute Simulation....PASS
gc_4x5_merra2_fullchem_complexSOA...................Execute Simulation....PASS
gc_4x5_merra2_fullchem_complexSOA_SVPOA.............Execute Simulation....PASS
gc_4x5_merra2_fullchem_LuoWd........................Execute Simulation....PASS
gc_4x5_merra2_fullchem_marinePOA....................Execute Simulation....PASS
gc_4x5_merra2_fullchem_RRTMG........................Execute Simulation....PASS
gc_4x5_merra2_Hg....................................Execute Simulation....PASS
gc_4x5_merra2_metals................................Execute Simulation....PASS
gc_4x5_merra2_POPs_BaP..............................Execute Simulation....PASS
gc_4x5_merra2_tagCH4................................Execute Simulation....PASS
gc_4x5_merra2_tagCO.................................Execute Simulation....PASS
gc_4x5_merra2_tagO3.................................Execute Simulation....PASS
gc_4x5_merra2_TransportTracers......................Execute Simulation....PASS
gc_4x5_merra2_TransportTracers_LuoWd................Execute Simulation....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 26
Execution tests failed: 0
Execution tests not yet completed: 0

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%  All execution tests passed!  %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

**Comment by yantosca** [comment]
GCHP integration tests all passed except tagO3.  This is a known issue. We will work on fixing this shortly, but it should not hold up the merging of this PR.
```console
==============================================================================
GCHP: Execution Test Results

GCClassic #d9afaf7 HEMCO submod update: Merge PR #191 into HEMCO 3.6.1
GEOS-Chem #54b11c61d Stage code for PR #1583 for merging into GEOS-Chem 14.1.1
HEMCO     #cf7d4ed GEOS-only updates by Christoph Keller

Number of execution tests: 5

Submitted as SLURM job: 43293748
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gchp_merra2_fullchem................................Execute Simulation....PASS
gchp_merra2_fullchem_benchmark......................Execute Simulation....PASS
gchp_merra2_fullchem_RRTMG..........................Execute Simulation....PASS
gchp_merra2_tagO3...................................Execute Simulation....FAIL
gchp_merra2_TransportTracers........................Execute Simulation....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 4
Execution tests failed: 1
Execution tests not yet completed: 0
```

**Pull request merged**